File: LeetCode/python3/714.py
class Solution:
    def maxProfit(self, prices, fee):
        """
        :type prices: List[int]
        :type fee: int
        :rtype: int
        """
        size = len(prices)
        if size < 2:
            return 0
        profit = 0
        cost = prices[0]
        for i in range(1, size):
            if prices[i] < cost:
                cost = prices[i]
            elif prices[i] - fee > cost:
                profit += prices[i] - fee - cost
                cost = prices[i] - fee
        return profit

File: LeetCode/cpp/714.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices, int fee) {
        auto size = prices.size();
        if (size == 0)
            return 0;
        vector<int> hold(size, 0), notHold(size, 0);
        hold[0] = -prices[0];
        for (int i = 1; i < size; ++i) {
            hold[i] = max(hold[i - 1], notHold[i - 1] - prices[i]);
            notHold[i] = max(notHold[i - 1], hold[i - 1] + prices[i] - fee);
        }
        return notHold[size - 1];
    }
};
